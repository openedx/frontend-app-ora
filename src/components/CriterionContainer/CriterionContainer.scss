@import "~@edx/brand/paragon/variables";
@import "~@edx/paragon/scss/core/core";
@import "~@edx/brand/paragon/overrides";

.criteria-label {
  width: 100%;
  .criteria-title {
    display: inline-block;
    max-width: calc(100% - 44px);
    color: $primary-500;
    font-weight: bold;
    vertical-align: top;
  }
  .esg-help-icon {
    margin-top: (map-get($spacers, 1\.5) * -1);
    margin-right: (map-get($spacers, 2\.5) * -1);
    vertical-align: top;
  }
}
.criteria-option {
  width: 100%;
  > div {
    display: inline;
    width: 100%;
    .pgn__form-label {
      display: inline-flex;
    }
    .pgn__form-control-description {
      float: right;
    }
  }
}

.criterion-feedback {
  margin-top: 1rem;
}
