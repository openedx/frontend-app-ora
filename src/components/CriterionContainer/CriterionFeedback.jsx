import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { useViewStep } from 'hooks/routing';
import { stepNames } from 'constants';
import {
  useShowValidation,
  useCriterionFeedbackFormFields,
} from 'hooks/assessment';

import messages from './messages';

/**
 * <CriterionFeedback />
 */
const CriterionFeedback = ({ criterion, criterionIndex }) => {
  console.log({ CriterionFeedback: { criterion, criterionIndex } });
  const { formatMessage } = useIntl();
  const step = useViewStep();
  const showValidation = useShowValidation();
  const { isInvalid, value, onChange } = useCriterionFeedbackFormFields(criterionIndex);

  if (step === stepNames.studentTraining) {
    return null;
  }

  const { feedbackEnabled, feedbackRequired } = criterion;
  if (!feedbackEnabled) {
    return null;
  }
  let commentMessage = formatMessage(messages.addComments);
  if (!feedbackRequired) {
    commentMessage += ` ${formatMessage(messages.optional)}`;
  }

  return (
    <Form.Group isInvalid={showValidation && isInvalid}>
      <Form.Control
        as="textarea"
        className="criterion-feedback feedback-input"
        floatingLabel={commentMessage}
        value={value}
        onChange={onChange}
      />
      {showValidation && isInvalid && (
        <Form.Control.Feedback type="invalid" className="feedback-error-msg">
          {formatMessage(messages.criterionFeedbackError)}
        </Form.Control.Feedback>
      )}
    </Form.Group>
  );
};

CriterionFeedback.propTypes = {
  criterion: PropTypes.shape({
    feedbackEnabled: PropTypes.bool.isRequired,
    feedbackRequired: PropTypes.bool.isRequired,
  }).isRequired,
  criterionIndex: PropTypes.number.isRequired,
};

export default CriterionFeedback;
