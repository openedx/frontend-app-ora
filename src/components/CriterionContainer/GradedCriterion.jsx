import React from 'react';
import PropTypes from 'prop-types';

import { Form, FormControlFeedback } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';

/**
 * <GradedCriterion />
 */
const GradedCriterion = ({ selectedOption, feedbackValue }) => (
  <div className="review-criterion">
    <Form.Label className="option-label">{selectedOption.name}</Form.Label>
    <div key={selectedOption.name} className="criteria-option">
      <div>
        <FormControlFeedback className="option-points">
          <FormattedMessage
            {...messages.optionPoints}
            values={{ points: selectedOption.points }}
          />
        </FormControlFeedback>
        {feedbackValue && <div className="rubric-feedback">{feedbackValue}</div>}
      </div>
    </div>
  </div>
);

GradedCriterion.defaultProps = {
  feedbackValue: null,
};
GradedCriterion.propTypes = {
  selectedOption: PropTypes.shape({
    name: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
  }).isRequired,
  feedbackValue: PropTypes.string,
};

export default GradedCriterion;
