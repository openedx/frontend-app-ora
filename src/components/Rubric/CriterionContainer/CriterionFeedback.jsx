import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import { feedbackRequirement } from 'data/services/lms/constants';

import messages from './messages';

/**
 * <CriterionFeedback />
 */
const CriterionFeedback = ({ criterion }) => {
  const { formatMessage } = useIntl();

  let commentMessage = formatMessage(messages.addComments);
  if (criterion.feedbackRequired === feedbackRequirement.optional) {
    commentMessage += ` ${formatMessage(messages.optional)}`;
  }

  const { feedbackValue, feedbackIsInvalid, feedbackOnChange } = criterion;

  if (
    !criterion.feedbackEnabled
    || criterion.feedbackRequired === feedbackRequirement.disabled
  ) {
    return null;
  }

  return (
    <Form.Group isInvalid={feedbackIsInvalid}>
      <Form.Control
        as="textarea"
        className="criterion-feedback feedback-input"
        floatingLabel={commentMessage}
        value={feedbackValue}
        onChange={feedbackOnChange}
      />
      {feedbackIsInvalid && (
        <Form.Control.Feedback type="invalid" className="feedback-error-msg">
          {formatMessage(messages.criterionFeedbackError)}
        </Form.Control.Feedback>
      )}
    </Form.Group>
  );
};

CriterionFeedback.propTypes = {
  criterion: PropTypes.shape({
    feedbackValue: PropTypes.string.isRequired,
    feedbackIsInvalid: PropTypes.bool.isRequired,
    feedbackOnChange: PropTypes.func.isRequired,
    feedbackEnabled: PropTypes.bool.isRequired,
    feedbackRequired: PropTypes.oneOf(Object.values(feedbackRequirement)).isRequired,
  }).isRequired,
};

export default CriterionFeedback;
