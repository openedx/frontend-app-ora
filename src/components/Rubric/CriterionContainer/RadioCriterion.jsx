import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import messages from './messages';

/**
 * <RadioCriterion />
 */
const RadioCriterion = ({ isGrading, criterion }) => {
  const { formatMessage } = useIntl();

  const { optionsValue, optionsIsInvalid, optionsOnChange } = criterion;

  return (
    <Form.RadioSet name={criterion.name} value={optionsValue}>
      {criterion.options.map((option) => (
        <Form.Radio
          className="criteria-option"
          key={option.name}
          value={option.name}
          description={formatMessage(messages.optionPoints, {
            points: option.points,
          })}
          onChange={optionsOnChange}
          disabled={!isGrading}
        >
          {option.name}
        </Form.Radio>
      ))}
      {optionsIsInvalid && (
        <Form.Control.Feedback type="invalid" className="feedback-error-msg">
          {formatMessage(messages.rubricSelectedError)}
        </Form.Control.Feedback>
      )}
    </Form.RadioSet>
  );
};

RadioCriterion.propTypes = {
  isGrading: PropTypes.bool.isRequired,
  criterion: PropTypes.shape({
    optionsValue: PropTypes.string.isRequired,
    optionsIsInvalid: PropTypes.bool.isRequired,
    optionsOnChange: PropTypes.func.isRequired,
    name: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        points: PropTypes.number.isRequired,
      }),
    ).isRequired,
  }).isRequired,
};

export default RadioCriterion;
