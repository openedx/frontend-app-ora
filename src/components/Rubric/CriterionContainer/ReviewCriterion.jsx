import React from 'react';
import PropTypes from 'prop-types';

import { Form, FormControlFeedback } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';

/**
 * <ReviewCriterion />
 */
const ReviewCriterion = ({ criterion }) => (
  <div className="review-criterion">
    {criterion.options.map((option) => (
      <div key={option.name} className="criteria-option">
        <div>
          <Form.Label className="option-label">{option.name}</Form.Label>
          <FormControlFeedback className="option-points">
            <FormattedMessage {...messages.optionPoints} values={{ points: option.points }} />
          </FormControlFeedback>
        </div>
      </div>
    ))}
  </div>
);

ReviewCriterion.propTypes = {
  criterion: PropTypes.shape({
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        points: PropTypes.number.isRequired,
      }),
    ).isRequired,
  }).isRequired,
};

export default ReviewCriterion;
