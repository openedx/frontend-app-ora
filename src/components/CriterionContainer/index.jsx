import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';

import InfoPopover from 'components/InfoPopover';

import './CriterionContainer.scss';

/**
 * <CriterionContainer />
 */
const CriterionContainer = ({
  criterion,
  input,
  feedback,
}) => (
  <Form.Group>
    <Form.Label className="criteria-label">
      <span className="criteria-title">{criterion.name}</span>
      <InfoPopover>
        <div className="help-popover-option">{criterion.description}</div>
        <hr />
        {criterion.options.map((option) => (
          <div key={option.name} className="help-popover-option">
            <strong>{option.label}</strong>
            <br />
            {option.description}
          </div>
        ))}
      </InfoPopover>
    </Form.Label>
    <div className="rubric-criteria">
      {input}
    </div>
    {feedback}
  </Form.Group>
);

CriterionContainer.defaultProps = {
  input: null,
  feedback: null,
};
CriterionContainer.propTypes = {
  input: PropTypes.node,
  feedback: PropTypes.node,
  criterion: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
      }),
    ).isRequired,
  }).isRequired,
};

export default CriterionContainer;
