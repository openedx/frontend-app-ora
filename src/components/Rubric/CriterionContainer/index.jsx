import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';

import InfoPopover from 'components/InfoPopover';
import RadioCriterion from './RadioCriterion';
import CriterionFeedback from './CriterionFeedback';
import ReviewCriterion from './ReviewCriterion';

/**
 * <CriterionContainer />
 */
const CriterionContainer = ({
  isGrading,
  criterion,
}) => (
  <Form.Group>
    <Form.Label className="criteria-label">
      <span className="criteria-title">{criterion.name}</span>
      <InfoPopover>
        <div className="help-popover-option">{criterion.description}</div>
        <hr />
        {criterion.options.map((option) => (
          <div key={option.name} className="help-popover-option">
            <strong>{option.name}</strong>
            <br />
            {option.description}
          </div>
        ))}
      </InfoPopover>
    </Form.Label>
    <div className="rubric-criteria">
      {isGrading ? (
        <RadioCriterion {...{ criterion, isGrading }} />
      ) : (
        <ReviewCriterion criterion={criterion} />
      )}
    </div>
    {isGrading && <CriterionFeedback {...{ criterion }} />}
  </Form.Group>
);

CriterionContainer.propTypes = {
  isGrading: PropTypes.bool.isRequired,
  criterion: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
      }),
    ).isRequired,
  }).isRequired,
};

export default CriterionContainer;
