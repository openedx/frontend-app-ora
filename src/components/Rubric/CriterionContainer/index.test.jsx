import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import CriterionContainer from '.';

jest.mock('./RadioCriterion', () => 'RadioCriterion');
jest.mock('./CriterionFeedback', () => 'CriterionFeedback');
jest.mock('./ReviewCriterion', () => 'ReviewCriterion');

describe('<CriterionContainer />', () => {
  const props = {
    isGrading: true,
    criterion: {
      name: 'criterion-1',
      description: 'description-1',
      options: [
        {
          name: 'option-1',
          description: 'description-1',
          points: 1,
        },
        {
          name: 'option-2',
          description: 'description-2',
          points: 2,
        },
      ],
    },
  };
  describe('renders', () => {
    test('is grading', () => {
      const wrapper = shallow(<CriterionContainer {...props} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('RadioCriterion')).toHaveLength(1);
      expect(wrapper.instance.findByType('ReviewCriterion')).toHaveLength(0);
      expect(wrapper.instance.findByType('CriterionFeedback')).toHaveLength(1);
    });

    test('is not grading', () => {
      const wrapper = shallow(<CriterionContainer {...props} isGrading={false} />);
      expect(wrapper.snapshot).toMatchSnapshot();

      expect(wrapper.instance.findByType('RadioCriterion')).toHaveLength(0);
      expect(wrapper.instance.findByType('ReviewCriterion')).toHaveLength(1);
      expect(wrapper.instance.findByType('CriterionFeedback')).toHaveLength(0);
    });
  });
});
