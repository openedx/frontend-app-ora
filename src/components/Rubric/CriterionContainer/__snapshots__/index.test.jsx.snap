// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CriterionContainer /> renders is grading 1`] = `
<Form.Group>
  <Form.Label
    className="criteria-label"
  >
    <span
      className="criteria-title"
    >
      criterion-1
    </span>
    <InfoPopover
      onClick={[Function]}
    >
      <div
        className="help-popover-option"
      >
        description-1
      </div>
      <hr />
      <div
        className="help-popover-option"
        key="option-1"
      >
        <strong>
          option-1
        </strong>
        <br />
        description-1
      </div>
      <div
        className="help-popover-option"
        key="option-2"
      >
        <strong>
          option-2
        </strong>
        <br />
        description-2
      </div>
    </InfoPopover>
  </Form.Label>
  <div
    className="rubric-criteria"
  >
    <RadioCriterion
      criterion={
        Object {
          "description": "description-1",
          "name": "criterion-1",
          "options": Array [
            Object {
              "description": "description-1",
              "name": "option-1",
              "points": 1,
            },
            Object {
              "description": "description-2",
              "name": "option-2",
              "points": 2,
            },
          ],
        }
      }
      isGrading={true}
    />
  </div>
  <CriterionFeedback
    criterion={
      Object {
        "description": "description-1",
        "name": "criterion-1",
        "options": Array [
          Object {
            "description": "description-1",
            "name": "option-1",
            "points": 1,
          },
          Object {
            "description": "description-2",
            "name": "option-2",
            "points": 2,
          },
        ],
      }
    }
  />
</Form.Group>
`;

exports[`<CriterionContainer /> renders is not grading 1`] = `
<Form.Group>
  <Form.Label
    className="criteria-label"
  >
    <span
      className="criteria-title"
    >
      criterion-1
    </span>
    <InfoPopover
      onClick={[Function]}
    >
      <div
        className="help-popover-option"
      >
        description-1
      </div>
      <hr />
      <div
        className="help-popover-option"
        key="option-1"
      >
        <strong>
          option-1
        </strong>
        <br />
        description-1
      </div>
      <div
        className="help-popover-option"
        key="option-2"
      >
        <strong>
          option-2
        </strong>
        <br />
        description-2
      </div>
    </InfoPopover>
  </Form.Label>
  <div
    className="rubric-criteria"
  >
    <ReviewCriterion
      criterion={
        Object {
          "description": "description-1",
          "name": "criterion-1",
          "options": Array [
            Object {
              "description": "description-1",
              "name": "option-1",
              "points": 1,
            },
            Object {
              "description": "description-2",
              "name": "option-2",
              "points": 2,
            },
          ],
        }
      }
    />
  </div>
</Form.Group>
`;
